contrib/babelfishpg_tsql/src/procedures.c

```diff
@@ -1283,146 +1287,299 @@ sp_describe_undeclared_parameters_internal(PG_FUNCTION_ARGS)
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  = -1 THEN \'varchar(max)\' "
+		"ELSE \'varchar(\' + CAST( CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  AS sys.VARCHAR(10) ) + \')\' "
 		"END "
-		"WHEN T2.name COLLATE sys.database_default = \'nvarchar\' THEN "
-		"CASE WHEN C.max_length = -1 THEN \'nvarchar(max)\' "
-		"ELSE \'nvarchar(\' + CAST( C.max_length/2 AS sys.VARCHAR(10) ) + \')\' "
+		"WHEN \'nvarchar\' THEN "
+		"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  = -1 THEN \'nvarchar(max)\' "
+		"ELSE \'nvarchar(\' + CAST( CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END /2 AS sys.VARCHAR(10) ) + \')\' "
 		"END "
-		"WHEN T2.name COLLATE sys.database_default = \'varbinary\' THEN "
-		"CASE WHEN C.max_length = -1 THEN \'varbinary(max)\' "
-		"ELSE \'varbinary(\' + CAST( C.max_length AS sys.VARCHAR(10) ) + \')\' "
+		"WHEN \'varbinary\' THEN "
+		"CASE WHEN CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  = -1 THEN \'varbinary(max)\' "
+		"ELSE \'varbinary(\' + CAST( CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  AS sys.VARCHAR(10) ) + \')\' "
 		"END "
 		"ELSE T2.name "
 		"END "					/* AS "suggested_system_type_name" */
-		", CASE "
-		"WHEN T2.name COLLATE sys.database_default IN (\'image\', \'ntext\',\'text\') THEN -1 "
-		"ELSE C.max_length "
+		", CASE T2.name COLLATE sys.database_default "
+		"WHEN \'image\' THEN -1 "
+		"WHEN \'ntext\' THEN -1 "
+		"WHEN \'text\' THEN -1 "
+		"ELSE CASE "
+				"WHEN a.atttypmod != -1 THEN "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, a.atttypmod) "
+				"ELSE "
+					"sys.tsql_type_max_length_helper(coalesce(tsql_type_name, tsql_base_type_name), a.attlen, t.typtypmod) "
+			"END  "
```

Comment:
sys.types view already contains max_length, precision and scale columns. Can we check if can reuse those instead of calculating it again and again.